crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs

```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+            let from_assign_to_loop = TextRange::new(binding_stmt.end(), for_stmt.start());
+            // Test if there's any reference to the list symbol between its definition and the for loop.
+            // if there's at least one, then it's been accessed in the middle somewhere, so it's not safe to change into a list comprehension
+            !binding
+                .references()
+                .map(|ref_id| checker.semantic().reference(ref_id).range())
+                .any(|text_range| from_assign_to_loop.contains_range(text_range))
+        });
+    // A dict update works in every context, while a dict comprehension only works when all the criteria are true
+    let fix_type = if binding_is_empty_dict
+        && assignment_in_same_statement
+        && binding_has_one_target
+        && binding_unused_between
+    {
+        DictComprehensionType::Comprehension
+    } else {
+        DictComprehensionType::Update
+    };
+
+    let mut diagnostic = Diagnostic::new(
+        ManualDictComprehension {
+            fix_type,
+            is_async: for_stmt.is_async,
+        },
+        *range,
+    );
+
+    if checker.settings.preview.is_enabled() {
+        // k: v inside the comprehension
+        let to_append = (slice.as_ref(), value.as_ref());
+
+        diagnostic.set_fix(convert_to_dict_comprehension(
+            fix_type,
+            binding,
+            for_stmt,
+            if_test.map(std::convert::AsRef::as_ref),
+            to_append,
+            checker,
+        ));
+    }
+    checker.report_diagnostic(diagnostic);
+}
+
+fn convert_to_dict_comprehension(
+    fix_type: DictComprehensionType,
+    binding: &Binding,
+    for_stmt: &ast::StmtFor,
+    if_test: Option<&ast::Expr>,
+    to_append: (&Expr, &Expr),
+    checker: &Checker,
+) -> Fix {
+    let locator = checker.locator();
+
+    let if_str = match if_test {
+        Some(test) => {
+            // If the test is an assignment expression,
+            // we must parenthesize it when it appears
+            // inside the comprehension to avoid a syntax error.
+            //
+            // Notice that we do not need `any_over_expr` here,
+            // since if the assignment expression appears
+            // internally (e.g. as an operand in a boolean
+            // operation) then it will already be parenthesized.
+            if test.is_named_expr() {
+                format!(" if ({})", locator.slice(test.range()))
+            } else {
+                format!(" if {}", locator.slice(test.range()))
+            }
+        }
+        None => String::new(),
+    };
+
+    // if the loop target was an implicit tuple, add parentheses around it
+    // ```python
+    //  for i in a, b:
+    //      ...
+    // ```
+    // becomes
+    // [... for i in (a, b)]
+    let iter_str = if for_stmt
+        .iter
+        .as_ref()
+        .as_tuple_expr()
+        .is_some_and(|expr| !expr.parenthesized)
```

Comment:
This is slightly shorter but maybe not more readable. Up to you.
```suggestion
    let iter_str = if let Expr::Tuple(ast::ExprTuple {
        parenthesized: false,
        ..
    }) = &*for_stmt.iter
```